### PR TITLE
cdc: add ticdc checksum verification sample

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.2 (TiDB 7.2 versions)
- [x] v7.1 (TiDB 7.1 versions)
- [ ] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by nongfushanquan — APPROVED** [review]

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 1]
```diff
@@ -0,0 +1,530 @@
+---
+title: TiCDC 行数据 Checksum 校验
+summary: TiCDC、Kafka、行数据 Checksum 校验
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:3]:
```suggestion
summary: 介绍 TiCDC 行数据 Checksum 校验的具体实现。
```

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+aliases: ['/zh/tidb/dev/ticdc-checksum-verification']
+---
+
+# TiCDC 行数据 Checksum 校验
+
+本文主要介绍如何消费 TiCDC 发送到 Kafka 的经由 Avro 协议编码的数据，以及[基于单行数据 Checksum 功能进行数据校验](https://docs.pingcap.com/zh/tidb/dev/ticdc-integrity-check)。
+
+本示例代码位于 [avro-checksum-verification](https://github.com/pingcap/tiflow/tree/master/examples/golang/avro-checksum-verification) 目录下。
+
+使用 [kafka-go](https://github.com/segmentio/kafka-go) 实现一个简单的 kafka consumer 程序，该程序不断地从指定的 Topic 读取数据，计算并且校验 Checksum。
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:13]:
```suggestion
本文主要介绍如何消费 TiCDC 发送到 Kafka 的经由 Avro 协议编码的数据，以及[基于单行数据 Checksum 功能进行数据校验](https://docs.pingcap.com/zh/tidb/dev/ticdc-integrity-check)。

本示例代码位于 [`avro-checksum-verification`](https://github.com/pingcap/tiflow/tree/master/examples/golang/avro-checksum-verification) 目录下。

本文使用 [kafka-go](https://github.com/segmentio/kafka-go) 实现一个简单的 Kafka consumer 程序。该程序不断地从指定的 Topic 中读取数据、计算并校验 Checksum 值。
```

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+    Name     string `json:"name"`
+    SchemaID int    `json:"id"`
+    Schema   string `json:"schema"`
+}
+
+```
+
+## 计算并校验 Checksum
+
+上一步拿到的 `valueMap` 和 `valueSchema`，即含有用于 checksum 计算和校验的所有元素。
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:212]:
```suggestion
上一步获取的 `valueMap` 和 `valueSchema` 包含了所有用于 Checksum 计算和校验的元素。
```

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+
+在消费端计算和校验 Checksum 的过程可以总结为以下几个步骤：
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:214]:
```suggestion
在消费端计算和校验 Checksum 的过程包含以下几个步骤：
```

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+
+1. schema 中含有每个列的类型信息，列顺序和 Checksum 计算顺序相同。
+2. 遍历每一列，根据列的数据值和对应的 MySQL Type，生成字节切片，不断更新 Checksum。
+3. 将上一步计算得到的 Checksum 和从收到的消息里提取出来的 Checksum 做比较。如果不一致，则说明 Checksum 校验失败，数据可能发生损坏。
+
+示例代码如下：
+
+```go
+func CalculateAndVerifyChecksum(valueMap, valueSchema map[string]interface{}) error {
+    // fields 存放有数据变更事件的每一个列的类型信息，按照每一列的 ID 排序，该顺序和 Checksum 计算顺序相同
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:224]: 其他的代码注释统一为中文

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+```
+
+## 计算并校验 Checksum
+
+上一步拿到的 `valueMap` 和 `valueSchema`，即含有用于 checksum 计算和校验的所有元素。
+
+在消费端计算和校验 Checksum 的过程可以总结为以下几个步骤：
+
+1. schema 中含有每个列的类型信息，列顺序和 Checksum 计算顺序相同。
+2. 遍历每一列，根据列的数据值和对应的 MySQL Type，生成字节切片，不断更新 Checksum。
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:217]:
```suggestion
2. 遍历每一列，根据列的数据值和对应的 MySQL Type，生成字节切片，不断更新 Checksum 值。
```

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+
+```
+
+## 计算并校验 Checksum
+
+上一步拿到的 `valueMap` 和 `valueSchema`，即含有用于 checksum 计算和校验的所有元素。
+
+在消费端计算和校验 Checksum 的过程可以总结为以下几个步骤：
+
+1. schema 中含有每个列的类型信息，列顺序和 Checksum 计算顺序相同。
```
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:216]:
```suggestion
1. 获取期望的 Checksum 值。
```
“schema 中含有每个列的类型信息，列顺序和 Checksum 计算顺序相同。”这个描述似乎不是实现的步骤，而是前提条件

File: develop/dev-guide-ticdc-checksum-verification.md
Comment (Oreoxmt) [inline develop/dev-guide-ticdc-checksum-verification.md:1]: 新增文档请添加到 TOC.md 中

File: ticdc/ticdc-avro-checksum-verification.md [hunk 8]
```diff
@@ -0,0 +1,518 @@
+---
+title: 基于 Avro 的 TiCDC 行数据 Checksum 校验
+summary: 介绍 TiCDC 行数据 Checksum 校验的具体实现。
+---
+
+# TiCDC 行数据 Checksum 校验
+
+本文主要介绍如何消费 TiCDC 发送到 Kafka 的经由 Avro 协议编码的数据，以及[基于单行数据 Checksum 功能进行数据校验](https://docs.pingcap.com/zh/tidb/dev/ticdc-integrity-check)。
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:8]:
```suggestion
本文主要介绍如何使用 Golang 消费 TiCDC 发送到 Kafka 的经由 Avro 协议编码的数据，以及如何基于[单行数据 Checksum 功能](/ticdc/ticdc-integrity-check.md)进行数据校验。
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+
+本示例代码位于 [`avro-checksum-verification`](https://github.com/pingcap/tiflow/tree/master/examples/golang/avro-checksum-verification) 目录下。
+
+本文使用 [kafka-go](https://github.com/segmentio/kafka-go) 实现一个简单的 Kafka consumer 程序。该程序不断地从指定的 Topic 中读取数据、计算并校验 Checksum 值。
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:12]:
```suggestion
本文使用 [kafka-go](https://github.com/segmentio/kafka-go) 实现一个简单的 Kafka Consumer 程序。该程序不断地从指定的 Topic 中读取数据、计算并校验 Checksum 值。
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+
+```go
+
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:15]:
```suggestion
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+        }
+
+        // 4. 数据消费成功，提交 offset
+        if err := consumer.CommitMessages(ctx, message); err != nil {
+            log.Error("commit kafka message failed", zap.Error(err))
+            break
+        }
+    }
+}
+
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:94]:
```suggestion
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+```
+
+从上面的代码可以看出，`getValueMapAndSchema` 和 `CalculateAndVerifyChecksum` 是计算 Checksum 的关键步骤，下面分别介绍这两个函数的实现。
+
+## 解码数据以及获取相应的 Schema
+
+`getValueMapAndSchema` 方法的主要作用是解码数据以及获取相应的 Schema，二者均以类型 `map[string]interface{}` 返回。
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:101]:
```suggestion
`getValueMapAndSchema()` 方法的主要作用是解码数据以及获取相应的 schema，二者均以类型 `map[string]interface{}` 返回。
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 13]
```diff
@@ -0,0 +1,518 @@
...
+
+```go
+
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:104]:
```suggestion
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 14]
```diff
@@ -0,0 +1,518 @@
...
+    }
+    pos, posMax := 0, len(bytes)-1
+    for ; pos < posMax; pos++ {
+        if bytes[pos] != 0 {
+            break
+        }
+    }
+    return bytes[pos:]
+}
+
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:517]:
```suggestion
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 15]
```diff
@@ -0,0 +1,518 @@
+---
+title: 基于 Avro 的 TiCDC 行数据 Checksum 校验
+summary: 介绍 TiCDC 行数据 Checksum 校验的具体实现。
+---
+
+# TiCDC 行数据 Checksum 校验
```
Comment (Oreoxmt) [inline ticdc/ticdc-avro-checksum-verification.md:6]:
```suggestion
# 基于 Avro 的 TiCDC 行数据 Checksum 校验
```

File: TOC.md [hunk 16]
```diff
@@ -546,7 +546,9 @@
         - [TiCDC CSV Protocol](/ticdc/ticdc-csv.md)
       - [TiCDC Open API v2](/ticdc/ticdc-open-api-v2.md)
       - [TiCDC Open API v1](/ticdc/ticdc-open-api.md)
-      - [Storage sink 消费程序编写指引](/ticdc/ticdc-storage-consumer-dev-guide.md)
+      - TiCDC 数据消费
+        - [TiCDC 行数据 Checksum 校验](/develop/dev-guide-ticdc-checksum-verification.md)
```
Comment (Oreoxmt) [inline TOC.md:550]:
```suggestion
        - [基于 Avro 的 TiCDC 行数据 Checksum 校验](/develop/dev-guide-ticdc-checksum-verification.md)
```

File: TOC.md [hunk 17]
```diff
@@ -546,7 +546,9 @@
         - [TiCDC CSV Protocol](/ticdc/ticdc-csv.md)
       - [TiCDC Open API v2](/ticdc/ticdc-open-api-v2.md)
       - [TiCDC Open API v1](/ticdc/ticdc-open-api.md)
-      - [Storage sink 消费程序编写指引](/ticdc/ticdc-storage-consumer-dev-guide.md)
+      - TiCDC 数据消费
+        - [基于 Avro 的 TiCDC 行数据 Checksum 校验](/develop/dev-guide-ticdc-checksum-verification.md)
```
Comment (Oreoxmt) [inline TOC.md:550]:
```suggestion
        - [基于 Avro 的 TiCDC 行数据 Checksum 校验](ticdc/ticdc-avro-checksum-verification.md)
```

File: ticdc/ticdc-integrity-check.md [hunk 18]
```diff
@@ -96,23 +96,7 @@ fn checksum(columns) {
...
-    4. 遍历完 `fields` 后，就获取了所有列的数据。对于 Delete 事件，将解码得到的列设置为 `PreColumns`；对于 Insert 和 Update 事件，将解码得到的列设置为 `Columns`。
-
-Checksum 计算和校验的过程如下：
-
-1. 调用 [`extractExpectedChecksum`](https://github.com/pingcap/tiflow/blob/eb04aecaf8e61f7f9d67597c2d2ef1f44583dd79/pkg/sink/codec/avro/decoder.go#L281) 方法，获取期望的 Checksum 值。如果该方法返回 `false`，则说明该事件不需要进行 Checksum 校验，因为上游并没有发送 Checksum。这可能发生在 TiCDC 开启了 Checksum 但 TiDB 没有开启该功能时，或者当前事件发生在 Checksum 校验功能开启之前等场景。
-2. 调用 [`calculateChecksum`](https://github.com/pingcap/tiflow/blob/eb04aecaf8e61f7f9d67597c2d2ef1f44583dd79/pkg/sink/codec/avro/decoder.go#L461) 方法，遍历之前重建出来的所有列。利用 [buildChecksumBytes](https://github.com/pingcap/tiflow/blob/eb04aecaf8e61f7f9d67597c2d2ef1f44583dd79/pkg/sink/codec/avro/decoder.go#L482) 方法将每一列的 value 和 MySQL Type 编码为一个字节切片，然后使用该字节切片更新 Checksum 值。
-3. 通过 [`verifyChecksum`](https://github.com/pingcap/tiflow/blob/eb04aecaf8e61f7f9d67597c2d2ef1f44583dd79/pkg/sink/codec/avro/decoder.go#L444) 方法，进行 Checksum 计算和校验。将步骤 2 计算的值与步骤 1 获取的期望值进行比较。如果不相等，则说明 Checksum 校验失败，数据可能存在损坏的情况。
+基于 Golang 的 Avro 数据消费和 Checksum 校验，可以参考 [TiCDC 行数据 Checksum 校验](/develop/dev-guide-ticdc-checksum-verification.md)。
```
Comment (Oreoxmt) [inline ticdc/ticdc-integrity-check.md:99]:
```suggestion
基于 Golang 的 Avro 数据消费和 Checksum 校验，可以参考 [TiCDC 行数据 Checksum 校验](/ticdc/ticdc-avro-checksum-verification.md)。
```

**Review by Oreoxmt — APPROVED** [review]

**Review by Oreoxmt — APPROVED** [review]

File: develop/dev-guide-ticdc-checksum-verification.md [hunk 19]
```diff
@@ -0,0 +1,530 @@
...
+    expectedChecksum, err := strconv.ParseUint(expected, 10, 64)
+    if err != nil {
+        return errors.Trace(err)
+    }
+
+    // 2. 遍历每一个 field，计算 checksum 值
+    var actualChecksum uint32
+    // buf 用来存储每次更新 checksum 时使用的字节切片
+    buf := make([]byte, 0)
+    for _, item := range fields {
```
Comment (zhangjinpeng1987) [inline develop/dev-guide-ticdc-checksum-verification.md]: The ordering of fields is fixed or not?

Comment (3AceShowHand) [reply develop/dev-guide-ticdc-checksum-verification.md]: yes, it's fixed, the same as the checksum calculation.

File: ticdc/ticdc-avro-checksum-verification.md [hunk 20]
```diff
@@ -0,0 +1,514 @@
+---
+title: 基于 Avro 的 TiCDC 行数据 Checksum 校验
+summary: 介绍 TiCDC 行数据 Checksum 校验的具体实现。
+---
+
+# 基于 Avro 的 TiCDC 行数据 Checksum 校验
+
+本文主要介绍如何使用 Golang 消费 TiCDC 发送到 Kafka 的经由 Avro 协议编码的数据，以及如何基于[单行数据 Checksum 功能](/ticdc/ticdc-integrity-check.md)进行数据校验。
```
Comment (hfxsd) [inline ticdc/ticdc-avro-checksum-verification.md:8]:
```suggestion
本文介绍如何使用 Golang 消费 TiCDC 发送到 Kafka、且由 Avro 协议编码的数据，以及如何基于[单行数据 Checksum 功能](/ticdc/ticdc-integrity-check.md)进行数据校验。
```

File: ticdc/ticdc-avro-checksum-verification.md [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+        }
+    }
+}
+```
+
+从上面的代码可以看出，`getValueMapAndSchema()` 和 `CalculateAndVerifyChecksum()` 是计算 Checksum 的关键步骤，下面分别介绍这两个函数的实现。
+
+## 解码数据以及获取相应的 Schema
+
+`getValueMapAndSchema()` 方法的主要作用是解码数据以及获取相应的 schema，二者均以类型 `map[string]interface{}` 返回。
```
Comment (hfxsd) [inline ticdc/ticdc-avro-checksum-verification.md:99]: 这里的 二者 是指谁？`getValueMapAndSchema()` 和 `CalculateAndVerifyChecksum()` 吗？建议明确一下。
```suggestion
`getValueMapAndSchema()` 方法的主要作用是解码数据以及获取相应的 schema，二者均以 `map[string]interface{}` 类型返回。
```

**Review by hfxsd — APPROVED** [review]

**Review by Oreoxmt — APPROVED** [review]

**Review by nongfushanquan — APPROVED** [review]

**Pull request merged**